### PR TITLE
[feat]: [CDS-79476]: Added Support For Env Create flow.

### PR DESCRIPTION
**Pull request by senjucanon2**
## Describe your changes:
Context: GitX for Environments
Change: Create Flow for Remote Environments is implemented in this PR.
Testing: Locally create of remote Yaml for environments was observed.

## Checklist
- [ ] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53128)

### CODE REVIEW

File: 127-cd-nextgen-entities/src/main/java/io/harness/repositories/environment/custom/EnvironmentRepositoryCustomImpl.java [hunk 1]
```diff
@@ -53,6 +65,34 @@ public Page<Environment> findAll(Criteria criteria, Pageable pageable) {
...
+    }
+
+    // trying to create remote entity with git simplification disabled
+    if (!cdGitXEnvironment.isNewGitXEnabled(environmentToSave.getAccountId(), environmentToSave.getOrgIdentifier(),
+            environmentToSave.getProjectIdentifier())) {
+      throw new InvalidRequestException(GitXUtils.getErrorMessageForGitSimplificationNotEnabled(
+          environmentToSave.getOrgIdentifier(), environmentToSave.getProjectIdentifier()));
+    }
+
+    addGitParamsToServiceEntity(environmentToSave, gitEntityInfo);
```
Comment (yogesh-chauhan) [inline 127-cd-nextgen-entities/src/main/java/io/harness/repositories/environment/custom/EnvironmentRepositoryCustomImpl.java:87]: is this the correct name for this method ?

**Review by senjucanon2 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53128)* status: 9 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/NextGenModule.java` line 1002 at r1](https://reviewable.io/reviews/harness/harness-core/53128#gh-1335655865:-NfFXn3sDOKLQjR6SoEv:b-8f1vaz) ([raw file](https://github.com/harness/harness-core/blob/fff9e7de0d5785da62be56d54c99d0e7c23774a0/120-ng-manager/src/main/java/io/harness/ng/NextGenModule.java#L1002)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

why can't we use the CDGitXService itself ?
CDGitXService is not meant for "Service" entity but its a regular service meant for CDGitX
</blockquote></details>

I have now used it, earlier i though there may be some custom methods specific to env and services respectively.
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/environment/services/impl/EnvironmentServiceImpl.java` line 196 at r1](https://reviewable.io/reviews/harness/harness-core/53128#gh-1335659794:-NfFZDWf6FjI0mPHmdgb:b-qaeewz) ([raw file](https://github.com/harness/harness-core/blob/fff9e7de0d5785da62be56d54c99d0e7c23774a0/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/environment/services/impl/EnvironmentServiceImpl.java#L196)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

Earlier, we were rethrowing all exceptions ( except DuplicateKey as is), now are wrapping all of them under InvalidRequestException. How does this change the behaviour of exception handling in "Environment" save API ?
</blockquote></details>

The duplicate field exception for an env already existing is still there , it is just checked before operations, and the other InvalidRequestExceptions are added to be in sync with Pipeline/Templates.
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/repositories/environment/custom/EnvironmentRepositoryCustom.java` line 18 at r1](https://reviewable.io/reviews/harness/harness-core/53128#gh-1335662007:-NfFY4C74Zsn5CnIXsfi:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/fff9e7de0d5785da62be56d54c99d0e7c23774a0/127-cd-nextgen-entities/src/main/java/io/harness/repositories/environment/custom/EnvironmentRepositoryCustom.java#L18)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

nit: new line for readability
</blockquote></details>

Done
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/repositories/environment/custom/EnvironmentRepositoryCustomImpl.java` line 81 at r1](https://reviewable.io/reviews/harness/harness-core/53128#gh-1335667173:-NfFY5X89SMSo159Rcr6:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/fff9e7de0d5785da62be56d54c99d0e7c23774a0/127-cd-nextgen-entities/src/main/java/io/harness/repositories/environment/custom/EnvironmentRepositoryCustomImpl.java#L81)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

this should be cdGitXService
</blockquote></details>

Done
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/repositories/environment/custom/EnvironmentRepositoryCustomImpl.java` line 87 at r1](https://reviewable.io/reviews/harness/harness-core/53128#gh-1335667441:-NfFYE6pEuxOHh9WxahF:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/fff9e7de0d5785da62be56d54c99d0e7c23774a0/127-cd-nextgen-entities/src/main/java/io/harness/repositories/environment/custom/EnvironmentRepositoryCustomImpl.java#L87)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

is this the correct name for this method ?
</blockquote></details>

Done.

**Comment by senjucanon2** [comment]
trigger uts

**Comment by senjucanon2** [comment]
trigger uts

**Review by hingerabhinav — DISMISSED** [review]
> Reviewed 4 of 9 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53128)* status: 9 file(s) to review. 1 change requests. 4. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/NextGenModule.java` line 9 at r2](https://reviewable.io/reviews/harness/harness-core/53128#-NfK0Ifh5Ztj6Dlkunm3:-NfK0Ifh5Ztj6Dlkunm4:bf5cqla) ([raw file](https://github.com/harness/harness-core/blob/c6fcfc02922165fb0b7db88e577f982abe929479/120-ng-manager/src/main/java/io/harness/ng/NextGenModule.java#L9)):*
> ```Java
> 
> package io.harness.ng;
> ```

nit: extra line

**Comment by senjucanon2** [comment]
trigger uts, messagecheck

**Comment by senjucanon2** [comment]
trigger uts

**Review by yogesh-chauhan — DISMISSED** [review]

**Comment by senjucanon2** [comment]
trigger uts

**Comment by senjucanon2** [comment]
trigger uts

**Review by yogesh-chauhan — APPROVED** [review]

**Review by yogesh-chauhan — APPROVED** [review]

**Review by hingerabhinav — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53128#-:-NfUISS06pMiMSlKH014:bnfp4nl)* status: 11 file(s) to review. 4. unresolved discussions.

**Comment by hingerabhinav** [comment]
trigger review

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/environment/mappers/EnvironmentMapper.java [hunk 2]
```diff
@@ -135,6 +140,37 @@ public EnvironmentResponseDTO writeDTO(Environment environment) {
         .tags(convertToMap(environment.getTags()))
         .version(environment.getVersion())
         .yaml(environment.getYaml())
+        .entityGitDetails(getEntityGitDetails(environment))
```
Comment (hingerabhinav) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/environment/mappers/EnvironmentMapper.java:143]: fallbackBranch is missing in the DTO

**Review by hingerabhinav — CHANGES_REQUESTED** [review]

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/environment/mappers/EnvironmentMapper.java (hunk 2 above)
Comment (hingerabhinav) [reply 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/environment/mappers/EnvironmentMapper.java:143]: Lets add in the subsequent PR

**Pull request merged**